Add create_type parameter to sqltypes.SchemaType

ovangle opened the pull request:
<!-- Provide a general summary of your proposed changes in the Title field above -->

Adds support for `create_type` to `sqlalchemy.sql.sqltypes.SchemaType`.

Fixes #10604 

### Description

I don't expect this PR to be accepted -- the changes I've made are far more extensive than would be strictly required to actually implement the issue as described, but  I believe them to be mostly positive and enhance readability and maintainabilty of the code surrounding DDL generation. 

Firstly, I refactored the types so that the `SchemaType` mixin is now a fully fledged `SchemaItem`. Secondly, I rewrote the DDL  generation so that `SchemaType`s DDL is generated by `SchemaGenerator` and `SchemaDropper` in a manner consistent with other schema items, rather than relying on table and metadata DDL events. 

The implementation is still a little bit of a WIP, I expect to be working on this for another week yet. There are a lot of tests to write and documentation to add, but I wanted to advertise the progress I've already made and that I am actively working on this issue and to hopefully gain in-principle approval for the changes I'm proposing. The implementation _should_ at least be the entire *existing* test suite (excluding mypy and lint), which, considering what has been reworked was an important milestone.

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

ovangle committed:
preliminary refactor:

Move dialect implementation checking to callsite.

The current implementation is awkwardly split between
passing `variant_mappings` to the listener when
declaring the listener and then checking whether
the listener is running on the correct implementation
when invoked.

Instead, include the checks in a wrapper method for the
listener.

- remove methods `SchemaType._is_impl_of_type` and
  `SchemaType._get_variants_for_set_table`
- Add method `SchemaType._run_only_on_mapped_variant`
  which calls a listener function only if the listener
  is running on the correct variant of the type in the
  current dialect.

  This could conceivably be moved to `TypeEngine`
- Add method `SchemaType._call_

ovangle committed:
Allow specifying a custom schema generator/dropper on dialects.

- Remove `dialect` argument from `SchemaGenerator` and `SchemaDropper`
  classes as it can be inferred from the `connection` parameter.
  Add deprecation warning if two positional arguments are provided
  to either class.
- Add `Dialect.ddl_generator`, `Dialect.ddl_dropper` class vars
  and set defaults to `SchemaGenerator`, `SchemaDropper` respectively
  on `DefaultDialect`
- `MetaData.create_all`, `MetaData.drop_all` should infer the
  schema generator/dropper from the connection
- `Table.create`, `Table.drop` should infer the schema generator/dropper
  from the dialect
- `Sequence.create`, `Sequence.drop should infer the generator/dropper
  from the dialect
- `Enum.create`, `Enum.drop` should infer the generator/dropper from
  the dialect
- Remove import of of NULLTYPE in sql.selectable from top level of
  module to remove circular import dependency between sql.sqltypes and
  sql.schema

ovangle committed:
refactor: SchemaType should be a SchemaItem

- `type_api.TypeEngineMixin` should be `visitors.Visitable` to
  ensure `Visitable` always appears last when linearizing
  `TypeEngine` subclasses.
  (Actually, Visitable should arguably be a metacls, but I'm
   not confident to make that kind of change)
- `metadata` is always set on `SchemaType` when a value is
  available
- Add internal `SchemaType._set_metadata` method to clean up
  event handling

ovangle committed:
Add `initial_target` argument to InvokeDDLBase.__init__

- Add `is_metadata_operation`, `is_table_operation`, `is_type_operation`
  to InvokeDDLBase, allowing type generation to discriminate behavior
  depending on what class of schema item is requested the generation.
- Add `InvokeDDLBase.invoke` method to traverse the generator/droppers
  initial target.

ovangle committed:
Add `kind` to `SchemaType`.

A `SchemaType` kind represents a namespace within DDL specification.
In the PG implementation, distinct types would correspond to different
values for `pg_type.typtype`. For most types, this would be the
`__visit_name__` of the type engine.

Allows unambiguous declaration of schema type instances which share
the same name, but have different DDL specifications

- Add `supports_schema_type_kinds` classvar to `Dialect`.
  `kind` represents a unique namespace inside the schema in which
  `SchemaType` instances can be declared.

  Usually set to a tuple, but can be set to a literal `*` to
  indicate that all kinds would be accepted by this dialect.

- Initialise `DefaultDialect.supports_schema_type_kinds` to ()
- Initialise `StrCompileDialect.supports_schema_type_kinds` to
  `'*'`

- Add `has_schema_type` method to `Dialect` for checks whether
  there is a specific declared schema type
- Add `get_schema_type_names` method to `Dialect` to get all
  schema type names available in the database

ovangle committed:
Support schema type generation in .sql.ddl

- Add `CreateEnum`, `DropEnum` types to `.sql.schema` and export via
  `.schema`

- Add stubs for `visit_create_enum`, `visit_drop_enum` to sql compiler

- Update `.sql.ddl.SchemaGenerator`, `.sql.ddl.SchemaDropper`
	- Add `visit_enum` method to both classes
	- Implement `_can_create_schema_type`, `_can_drop_schema_type`
	  respectively.
	- Update metadata visitation to traverse schema types which
   	  are shared by multiple tables
 	- Update table visitation to traverse schema types which are
 	  internal to the table definition
- Add `.sql.ddl.resolve_schema_type` which attempts to resolve any
  schema types associated with a column type. A schema type is
  considered associated with a type if:
  	- The column type is a custom type and the implementation
   	  is a schema type
	- The column type is an array of schema types
	- The column type has a mapped variant which is a schema type
	  in the provided dialect.

- `.sql.schema.SchemaType` should not use DDL `before_create`, `after_drop`
  events to generate DDL.
	- Add `SchemaType.kind` property to indicate the DDL namespace
 	  for instances of the type
	- Add `SchemaType._key` property to use as a key into the
	  `MetaData._types` registry.
	- Remove `_create_events` parameter from `SchemaType.__init__`
	- Remove `_on_table_create`, `_on_table_drop`,
	  `_on_metadata_create`, `_on_metadata_drop` methods from
     	  implementation
	- Update `_set_metadata` method to register the type with
	  `MetaData._types`, do not subscribe to `MetaData` DDL
	  events.
	- Update `_set_table` method to not subscribe to `Table`
	  DDL events
	- Add `_supports_create` to add additional implementation
	  dependent checks for whether any dialect-specific implementation
 	  of a `SchemaType` would support DDL specification.

- Add `_types` to `.sql.schema.Metadata`, to use as a registry
  for declared schema types attached to the metadata.

- Update `.sql.sqltypes.Enum`:
	- Implement `supports_create` from `SchemaType`
	- Remove `_create_events` from `__init__`
	- Add `create_type` parameter
- Update `.sql.sqltypes.Boolean`
	- Add `create_type` parameter to `__init__`
	  (only to support type adaption)

- (Temporarily) Comment out tests from test/sql/test_metadata.py which
  test for `SchemaType` listening to DDL events
- Fix test in test/sql/test_types.py which was expecting `dispatch`
  to be unaltered during adaption (and removed testing for changes to
  `"_create_event"` since that argument was removed from ctors which
  used it)

ovangle committed:
Do not use import statement from function scope.

Decorate `.sql.selectable.Select._scalar_type` with
`util.preload_module()` so that it does not use an import
statement from function scope.

Additionally, since the circular dependency between
`.sql.sqltypes` and `.sql.schema` has been resolved
(with `.sql.schema` always being loaded first), functions
decorated with `util.preload_module` in sqltypes can instead
use types exported from schema directly.

This includes
	- `Enum._set_table`
	- `Boolean._set_table`

ovangle committed:
Update postgresql dialect to reflect changes to schema types

- Update `.dialects.postgresql.base.PGCompiler`:
	- Rename `visit_create_enum_type` to `visit_create_enum`
	- Rename `visit_drop_enum_type` to `visit_drop_enum`

- Add `dialects.postgresql.base.PGDDLGenerator` as an extension
  of `.sql.ddl.SchemaGenerator` which can handle generation
  of PG domains
- Add `dialects.postgresql.base.PGDDLDropper` as an extension
  of `.sql.ddl.SchemaDropper` which can handle generation
  of PG domains

- Updated `.dialects.postgresq.base.PGDialect`:
	- Bound `PGDialect.ddl_generator` and `PGDialect.ddl_dropper`
  	  class vars to `PGDDLGenerator` and `PGDDLDropper` respectively.
	- Add support for 'enum', 'DOMAIN' schema type kinds.

	- Add implementation of `Dialect.has_schema_type`
	- Add implementation of `Dialect.get_schema_type_names`

	Also, a small refactor:
	- Update `Dialect._pg_type_filter_schema` to not assume
  	  that the query is already joined to `pg_catalog.pg_namespace`
  	  (like `_pg_class_filter_scope_schema`)
	- Update `Dialect._load_enums`, `Dialect._load_domains`
	  to reflect change to `_pg_type_filter_schema`

- `.dialects.postgresql.named_types.NamedType` is now an alias
  of `.sql.schema.SchemaType`

- Removed implemetations of :
	- `.dialects.postgresq.named_types.NamedTypeGenerator`
  	- `.dialects.postgresql.named_types.NamedTypeDropper`
	- `.dialects.postgresql.named_types.EnumGenerator`
	- `.dialects.postgresql.named_types.EnumDropper`
	- `.dialects.postgresql.named_types.DomainGenerator`
	- `.dialects.postgresql.named_types.EnumDropper`
  (as their functionality is superseded by `PGDDLGenerator`,
  `PGDDLDropper`)

- Updated `.dialects.postgresql.named_types.ENUM` to respect
  changes to base class and removed unecessary overrides
  of `create` and `drop` methods

- Updated `.dialects.postgresql.named_types.DOMAIN` to respect
  changes to base class and removed unecessary overrides
  of `create` and `drop` methods

- `.dialects.postgresql.named_types.CreateEnumType` and
  `.dialects.postgresql.named_types.DropEnumType` are aliases of
  `.sql.ddl.CreateEnum`, `.sql.ddl.DropEnum` respectively.

- (temporarily) comment out tests relating to DDL events which are
  no longer subscribed to by ENUM, DOMAIN.
  These tests will be removed eventually
- Add skip to one test which cannot be re-implemented after refactor
  due to a small change in semantics

zzzeek commented:
OK, so I sort of see what's happening here, but this is the important part

what is the problem that hopes to be solved here?      

if readability alone, it's not clear this is helping, this looks a lot more complicated.  Especially that it removes postgresql.NamedType.   The "named type" concept is entirely a PostgreSQL thing right now, there's no other database I'm familiar with (PG variants excluded) that has this concept of CREATE TYPE.

The other reason SchemaType is around is to support types that also carry along a CHECK constraint with them, that is, the Boolean and base Enum datatypes.

That latter concept has proven to be mostly a disaster, particularly on the Alembic side, which is why the "create_constraint" parameter of Boolean/Enum now defaults to False.   Once we defaulted these to False, the user reported issues we we had with these features went to zero, and I'm pretty sure nobody is turning on these parameters.

this is all to say that SchemaType's movement is just as likely (if not moreso) to be **out** of SQLAlchemy entirely, rather than more *in* to SQLAlchemy.       If I'm looking to make a big change it could be to remove SchemaType entirely.  It's been the source of confusion and bugs for years.   

So I'm pretty skeptical about a big refactor making SchemaType more of a first class citizen without a really good reason.    IMO if we are looking for readability, removing SchemaType entirely for like 2.2 and just having PG stick to its NamedType thing would improve readability the most.

ovangle committed:
Fixes from self-review on github

- `SchemaDropper.visit_table` should use `resolve_schema_type`
- Some code in `Column.__init__` was somehow accidentally indented
- Do not delete value in `MetaData._types` registry unless the
  schema has actually changed

CaselIT commented:
@ovangle not sure if it works for you time wise, given Australia from the profile, but we do have dev meetings at https://app.gitter.im/#/room/#sqlalchemy_devel:gitter.im on Thursdays at 15 utc time, if you want to participate.
